### ISSUE TITLE
Separate data from code

### ISSUE THREAD

**Issue by ahangarha** [ahangarha/MV-Projects#26]
Host data in a separate location out of this repo.

Load the data from that location using FetchAPI.

**Comment by ahmetbozaci** [comment]
I can create API with ruby on rails and upload it to Heroku. Then we can use this link to send and get data.
But firstly ı have to create an ERD schema for the tables. It will be good if you create an empty repository for the backend side so I can work on it. Also, other people can develop the backend side.

**Comment by ahmetbozaci** [comment]
**For now, it can be like that. But in the future, we can add an email and password to the user.**

- Lucidchart [link](https://lucid.app/lucidchart/55b92360-362a-4817-8006-ff451cf93458/edit?invitationId=inv_95dfe4b0-5482-4e81-bb91-cc90ce3563e5) for edit

![image](https://user-images.githubusercontent.com/38635158/161751886-2558013e-e538-48b7-8271-0d8e579c881f.png)

**Comment by ahangarha** [comment]
> I can create API with ruby on rails and upload it to Heroku.

It would be great if you do this. I am just worried about who can update the data (because we don't want some wrong person to come and put wrong information.

For this matter and also the structure of the database and API we need to discuss.

**Comment by ahmetbozaci** [comment]
`Add` a button in the navigation bar opens the form. Only users can see this button. But we need a signup, login page for that.
For now, I can just create a basic API like in the schema, we can use it. In the future, we can change the API link to make it secure.

**Comment by ahangarha** [comment]
It would be great if you can replicate the current json in the API but then we can sit and discuss about challenges.

For example, not all projects are done by one person, so we need to have an array for list of authors.

I think we can work on it later. First step would be to "separate data from code" then will improve it. Like API v2 :)

**Comment by ahmetbozaci** [comment]
Yes, we can open new issues for new features. But ı need another repository to create a ruby on rails. You can create or ı can create?

**Comment by ahangarha** [comment]
You can create it. I don't mind. You are going to manage that. I still don't know RoR. So you should maintain it. Better you make it.

**Comment by ahmetbozaci** [comment]
I finish the database
I didn't change the link name. For now, we can use it. But in the future we can change also people can use this link in the env file so no one sees the link in the repository.

`baseURL = https://salty-badlands-19891.herokuapp.com/api/v1`
### GET METHODS
1. show all users: `baseURL/users` 
  **OUTPUT:**
![image](https://user-images.githubusercontent.com/38635158/162219826-6106df5f-6d45-4dbb-bf52-6dfa27036a6a.png)

2. show specific user's projects:`baseURL/users/id/projects`
**OUTPUT:**
![image](https://user-images.githubusercontent.com/38635158/162219935-4bd1cc10-2ec0-4e27-9684-6a9d5e0efebf.png)

3. show specific user's specific project: `baseURL/users/id/projects/id`
**OUTPUT:**
![image](https://user-images.githubusercontent.com/38635158/162219935-4bd1cc10-2ec0-4e27-9684-6a9d5e0efebf.png)

### POST METHODS
1. create user: `baseURL/signup`
_Name has to be unique._
_Picture has to be link for now in future if we use some cloud service it can be good_
**OUTPUT:**
![image](https://user-images.githubusercontent.com/38635158/162220973-e30e0d3b-4598-476e-bb3d-2cdae5577ba6.png)

2. create project: `baseURL/users/1/projects`
**OUTPUT:**
![image](https://user-images.githubusercontent.com/38635158/162222695-47dca856-0bc7-4315-8e5b-362573bc019b.png)

**Comment by ahangarha** [comment]
Great :+1: 

Just few suggestions:
- We need a `/projects` endpoint to get all the projects
- Please don't make post request open to everyone. I don't know if it is possible or not but it would be great if you make a token for authentication which remain hidden. This way, only authorized people can add/update data.
- For users, we can have more fields like their github account
- We need to have endpoint for `DELETE` and `PUT` methods as well.

**Comment by ahangarha** [comment]
Next week I have my solo capstone project for react/redux and I think I would finish it in 2-3 days. After that I have one week holiday. I think that time is a good time to migrate this project to use react and new API.

What do you think?

**Comment by ahmetbozaci** [comment]
> Great +1
> 
> Just few suggestions:
> 
> * We need a `/projects` endpoint to get all the projects
> * Please don't make post request open to everyone. I don't know if it is possible or not but it would be great if you make a token for authentication which remain hidden. This way, only authorized people can add/update data.
> * For users, we can have more fields like their github account
> * We need to have endpoint for `DELETE` and `PUT` methods as well.

- Ok ı can change endpoints. Make it more clear.
- Delete and put endpoints for the only project? Do we need them for users? Maybe no need to delete users but we can add the put method. 
- The user table will have password and email fields in the future these are for signup. I can create another table for that. Like profile table. This table can have a GitHub account, portfolio link maybe other social media links(Twitter, LinkedIn, medium, etc..)
- About the post request: Everyone can see the projects on the page but only users will see the add post button. We need to authenticate on react side. 

- I think we can add more tables. Like project category table.
- About the project picture: I think ıt will be good if we keep it as an URL. It can be easier for users. They just copy the image links from the project repository. No need to download and upload it. Also, cloud services have some storage limits.

I have time. I am in the job search module. I will update the rails part as ı can.

**Comment by ahangarha** [comment]
> Delete and put endpoints for the only project?

Actually this is were we collect projects. Users get extracted from projects. We don't add them manually. This is what I think at the moment. Any mechanism for generating users table from projects would be another issue. Would be good but not critical. I would say even we can generate users list on the fly.

> The user table will have password and email fields in the future these are for signup. I can create another table for that. Like profile table. This table can have a GitHub account, portfolio link maybe other social media links(Twitter, LinkedIn, medium, etc..)

Would be interesting for future. For that we might need to talk. I would like to see your perspective.

> I think we can add more tables. Like project category table.
I think I have already made a list of such tags (rather categories); something like technologies, module,...

> About the project picture: I think ıt will be good if we keep it as an URL
Exactly. We could even do some web-scrapping to find the screenshot image from the repo (in readme file) but I think we can ask the user to provide that link. We can explore other solutions. For example I think on front-end, as soon as we get the repo URL, we can make simple process and if we can detect image, automatically load it. User can choose to accept that image or add another URL.

These are all amazing features. Would be fun to implement them.
